core/src/cluster_info.rs

```diff
@@ -913,27 +913,35 @@ impl ClusterInfo {
     /// each Vec is no larger than `PROTOCOL_PAYLOAD_SIZE`
     /// Note: some messages cannot be contained within that size so in the worst case this returns
     /// N nested Vecs with 1 item each.
-    fn split_gossip_messages(mut msgs: Vec<CrdsValue>) -> Vec<Vec<CrdsValue>> {
+    fn split_gossip_messages(msgs: Vec<CrdsValue>) -> Vec<Vec<CrdsValue>> {
         let mut messages = vec![];
-        while !msgs.is_empty() {
-            let mut payload = vec![];
-            let mut size = serialized_size(&payload).expect("Couldn't check size");
-            while let Some(msg) = msgs.pop() {
-                let msg_size = msg.size();
-                if size + msg_size > MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
-                    if msg_size < MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
-                        msgs.push(msg);
-                    } else {
-                        debug!(
-                            "dropping message larger than the maximum payload size {:?}",
-                            msg
-                        );
+        let mut payload = vec![];
+        let base_size = serialized_size(&payload).expect("Couldn't check size");
+        let mut size = base_size;
+        for msg in msgs {
+            let msg_size = msg.size();
+            // If the message is too big to fit in this batch
+            if size + msg_size > MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
+                // See if it can fit in the next batch
+                if base_size + msg_size <= MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
```

Comment:
aren't you counting base_size twice?

how about

max = MAX_PROTOCOL_PAYLOAD_SIZE - base_size;
size = 0;

Reply:
Am I? Where? 
msg_size is the size of the crdsvalue itself.
Size is init to base size (which like you suggested I can move to rhs).
Size is then incremented by msg_size each time a new msg is added. 
When we move to a new transmission, size is reset to base size + rollover msg size and payload is init to carry the rollover message.